### PR TITLE
[SPARK-37473][CORE] BypassMergeSortShuffleWriter may loss data when disk is missing however catagory is present

### PR DESCRIPTION
**Pull request by seayoun**
### What changes were proposed in this pull request?
We think it has no data when the segment file not exists when all segment files produced by `BypassMergeSortShuffleWriter` is merging;

However, `file.exists()` may rerurn `false` when then the disk which segment file in on is missing and the root catagory exists; the missing disk only lead `file.exists()` return `false` but no exception. The task will run in peace without current segment file written.

The segment data will be ignored  and leading shuffle data loss.

### Why are the changes needed?
Shuffle may loss partition data leading corretness problems in final data.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
 the task finished as normal without exception when I delete  a partition file before segment files merge， and the final data is incorrect;
The task will retry when apply this patch

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Mind filling the PR description for:
- Does this PR introduce any user-facing change?
- How was this patch tested?

**Comment by seayoun** [comment]
ok to test

**Comment by seayoun** [comment]
> Mind filling the PR description for:
> 
> * Does this PR introduce any user-facing change?
> * How was this patch tested?

done

**Comment by LuciferYang** [comment]
> the task finished as normal without exception when I delete a partition file before segment files merge， and the final data is incorrect;
> The task will retry when apply this patch

I think we should add a new UT for this scene

**Review by Ngone51 — COMMENTED** [review]
Did you hit the issue in a production environment or else?  And if this's the case, I think we should also fix the merging on spill files, which could have the same problem.

**Comment by Ngone51** [comment]
cc @mridulm @tgravescs @attilapiros

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 1]
```diff
@@ -201,29 +206,36 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
    *
    * @return array of lengths, in bytes, of each partition of the file (used by map output tracker).
    */
-  private long[] writePartitionedData(ShuffleMapOutputWriter mapOutputWriter) throws IOException {
+  private long[] writePartitionedData(
+          ShuffleMapOutputWriter mapOutputWriter,
+          Set<Integer> nonEmptyPartitionSet) throws IOException {
```
Comment (tgravescs) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:211]: should add description of parameters and why we need the nonEmptyPartitionSet

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 2]
```diff
@@ -201,29 +206,36 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
...
+              if (!file.delete()) {
+                logger.error("Unable to delete file for partition {}", i);
+              }
             } else {
-              writePartitionedDataWithStream(file, writer);
-            }
-            if (!file.delete()) {
-              logger.error("Unable to delete file for partition {}", i);
+              throw new IOException(
+                      "Segment file " + file.getAbsolutePath() + " for partition " + i + " doesn't exists.");
```
Comment (tgravescs) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:238]: might be nice to add more description to the exception for user to understand what might have happened

**Review by tgravescs — COMMENTED** [review]
As others have asked, It would be great to know how this file was deleted?  Was there some kind of disk corruption going on though would seem weird just one file affected, did it actually occur in production?

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 3]
```diff
@@ -165,10 +167,13 @@ public void write(Iterator<Product2<K, V>> records) throws IOException {
...
       writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
 
+      Set<Integer> nonEmptyPartitionSet = new HashSet<>();
       while (records.hasNext()) {
         final Product2<K, V> record = records.next();
         final K key = record._1();
-        partitionWriters[partitioner.getPartition(key)].write(key, record._2());
+        int partitionId = partitioner.getPartition(key);
+        partitionWriters[partitionId].write(key, record._2());
+        nonEmptyPartitionSet.add(partitionId);
```
Comment (mridulm) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java:176]: This is updating the Set for every tuple processed.
Instead, why not use `partitionWriterSegments[i].length > 0` instead in `writePartitionedData` ?

**Comment by mridulm** [comment]
I was quite confused by the description.
What do we mean by `root category` ?

**Comment by Ngone51** [comment]
@seayoun could you update the PR?

**Pull request closed without merging**